File: lib/api/models/ai_training_text.dart
// lib/models/ai_training_text.dart

class AITrainingText {
  final int? textId;
  final String content;
  final String contentType;
  final int? categoryId;
  final String? description;
  final int? createdByUserId;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final int usageCount;
  final double? effectivenessRating;
  final bool isActive;

  // Relationships
  final Map<String, dynamic>? user;
  final Map<String, dynamic>? category;

  AITrainingText({
    this.textId,
    required this.content,
    required this.contentType,
    this.categoryId,
    this.description,
    this.createdByUserId,
    this.createdAt,
    this.updatedAt,
    this.usageCount = 0,
    this.effectivenessRating,
    this.isActive = true,
    this.user,
    this.category,
  });

  factory AITrainingText.fromJson(Map<String, dynamic> json) {
    return AITrainingText(
      textId: json['text_id'],
      content: json['content'],
      contentType: json['content_type'],
      categoryId: json['category_id'],
      description: json['description'],
      createdByUserId: json['created_by_user_id'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
      usageCount: json['usage_count'] ?? 0,
      effectivenessRating: json['effectiveness_rating'] != null
          ? (json['effectiveness_rating'] as num).toDouble()
          : null,
      isActive: json['is_active'] ?? true,
      user:
          json['user'] != null ? Map<String, dynamic>.from(json['user']) : null,
      category: json['category'] != null
          ? Map<String, dynamic>.from(json['category'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (textId != null) 'text_id': textId,
      'content': content,
      'content_type': contentType,
      if (categoryId != null) 'category_id': categoryId,
      if (description != null) 'description': description,
      if (createdByUserId != null) 'created_by_user_id': createdByUserId,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
      'usage_count': usageCount,
      if (effectivenessRating != null)
        'effectiveness_rating': effectivenessRating,
      'is_active': isActive,
      if (user != null) 'user': user,
      if (category != null) 'category': category,
    };
  }

  AITrainingText copyWith({
    int? textId,
    String? content,
    String? contentType,
    int? categoryId,
    String? description,
    int? createdByUserId,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? usageCount,
    double? effectivenessRating,
    bool? isActive,
    Map<String, dynamic>? user,
    Map<String, dynamic>? category,
  }) {
    return AITrainingText(
      textId: textId ?? this.textId,
      content: content ?? this.content,
      contentType: contentType ?? this.contentType,
      categoryId: categoryId ?? this.categoryId,
      description: description ?? this.description,
      createdByUserId: createdByUserId ?? this.createdByUserId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      usageCount: usageCount ?? this.usageCount,
      effectivenessRating: effectivenessRating ?? this.effectivenessRating,
      isActive: isActive ?? this.isActive,
      user: user ?? this.user,
      category: category ?? this.category,
    );
  }
}

File: final_project/models/ai_training_text.py
from datetime import datetime
from enum import Enum

from sqlalchemy import (
    Column, Integer, String, Text, Boolean, DateTime,
    Numeric, ForeignKey, CheckConstraint, Index, event
)
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from ..db import Base

# Enum for AI training text content types
class ContentType(str, Enum):
    HABIT_DESCRIPTION = "habit_description"
    TASK_EXAMPLE = "task_example"
    BEST_PRACTICE = "best_practice"

class AITrainingTextDB(Base):
    __tablename__ = "ai_training_text"

    text_id = Column(Integer, primary_key=True, autoincrement=True)
    content = Column(Text, nullable=False)
    content_type = Column(String(50), nullable=False)
    category_id = Column(Integer, ForeignKey("category.category_id", ondelete="SET NULL"), nullable=True)
    description = Column(Text, nullable=True)
    created_by_user_id = Column(Integer, ForeignKey("users.user_id", ondelete="SET NULL"), nullable=True)
    created_at = Column(DateTime, server_default=func.now(), nullable=False)
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now(), nullable=False)
    usage_count = Column(Integer, default=0, nullable=False)
    effectiveness_rating = Column(Numeric(3, 2), nullable=True)
    is_active = Column(Boolean, default=True, nullable=False)

    __table_args__ = (
        CheckConstraint(
            "content_type IN ('habit_description', 'task_example', 'best_practice')",
            name="check_content_type"
        ),
        Index("idx_ai_training_text_content_type", "content_type"),
        Index("idx_ai_training_text_category_id", "category_id"),
        Index("idx_ai_training_text_is_active", "is_active"),
    )

    # Relationships
  
    user = relationship(
        "UserDB",
        back_populates="ai_training_texts",
        foreign_keys=[created_by_user_id]
    )

    category = relationship("CategoryDB", back_populates="training_texts")

# Event listener to update 'updated_at' before updates
@event.listens_for(AITrainingTextDB, 'before_update')
def _update_timestamp(mapper, connection, target):
    target.updated_at = datetime.utcnow()
